src/middlewares/authMiddleware.js

```diff
@@ -0,0 +1,66 @@
+import jwt from 'jsonwebtoken';
+import { envConfig } from '../configs/env.js';
+import User from '../models/UserModel.js';
+
+// User authentication middleware
+export const isAuthenticatedUser = async (req, res, next) => {
+  try {
+    // Check if JWT exists in the "Authorization" header and extract it
+    const authHeader = req.header('Authorization');
+    const token = authHeader && authHeader.split(' ')[1];
+
+    // Check if no token is provided or not
+    if (!token) {
+      // If no token, send a 401 Unauthorized response
+      return res.status(401).json({ error: 'Unauthorized access. Token not provided.' });
+    }
+
+    try {
+      // Verify the JWT token
+      const user = await jwt.verify(token, envConfig.jwtSecret);
+
+      // Set the decoded user information in the request object
+      req.user = user;
+
+      // Call the next middleware or route handler
+      next();
+    } catch (jwtError) {
+      // If verification fails, send a 403 Forbidden response
+      return res.status(403).json({ error: 'Access forbidden. Invalid token.' });
+    }
+  } catch (error) {
+    // Handle unexpected errors
+    // eslint-disable-next-line no-console
+    console.error('Authentication middleware error:', error);
+    res.status(500).json({ error: 'Internal Server Error' });
+  }
+};
+
+// Admin authorization middleware
+export const isAdmin = async (req, res, next) => {
+  try {
```

Comment:
I think you haven't got my point here. In isAdmin middleware, it just simple is:

```
if (req.user && req.user.role === 'admin') {
    next();
  } else {
    res.status(httpStatus.FORBIDDEN).json({ message: 'Not authorised.' });
  }
```

Reply:
I want to separate the check of user and user role to make sure when checking for the existence of the user object after querying the database, we prevent errors occurred when the findById query does not find a user, what if it tried to access trying to access properties on a potentially null or undefined object?

Reply:
First, we need to determine whether the user is authenticated or not, and we achieve this by using the `isAuthenticatedUser` middleware. In our system, both regular users and administrators are considered users, correct? If the findById query does not return a user, it will already raise an alert indicating that the user does not exist and halt the process.

Now, if a user is found, we proceed to use the `isAdmin` middleware to ascertain whether the user has administrative privileges. Since we've already confirmed that this user is authenticated, we don't need to recheck their authentication status. Instead, we simply check if their `role` is marked as 'admin'. If the user's role is indeed 'admin', the process can continue. However, if it's not 'admin', we raise an alert indicating that access is forbidden, and the process cannot proceed.

Reply:
Again, I do understand what you meant and you're right when you say "If the findById query does not return a user, it will already raise an alert indicating that the user does not exist and halt the process." because there's already try catch to catch errors. But I want the error to be more specific and that it gets send with the message set in the response, that's why I separate it. 

If just because you want the code to be shorter, that's fine I can combine as if `(user && user.role === "admin") ` as I did in the first place.

Reply:
I would appreciate a more concise code to eliminate unnecessary complexity. Additionally, if you wish to enhance error specificity, it can be easily achieved by modifying the error message.

Reply:
Can I ask what do you mean by "modifying the error message."? 

would you be able to send a code snippets here of how you will refactor this to be more concise?